File: modules/lambda/python/src/gamelift_handler.py
"""
AWS Lambda Handler for GameLift Operations
This handler implements GameLift ListFleets API using boto3
"""

import json
import os
from typing import Dict, Any, Optional
import boto3
from botocore.exceptions import ClientError


def lambda_handler(event: Dict[str, Any], context: Any) -> Dict[str, Any]:
    """
    Main Lambda handler for GameLift operations
    
    Args:
        event: API Gateway event containing request information
        context: Lambda context object
        
    Returns:
        API Gateway response dictionary
    """
    
    # Get environment variables
    environment = os.environ.get('ENVIRONMENT', 'unknown')
    project = os.environ.get('PROJECT', 'unknown')
    
    # Parse the event
    http_method = event.get('httpMethod', 'GET')
    path = event.get('path', '/')
    
    # Initialize GameLift client
    try:
        gamelift_client = boto3.client('gamelift')
        aws_region = os.environ.get('AWS_REGION', 'us-east-1')
    except Exception as e:
        return create_error_response(
            500,
            "Failed to initialize GameLift client",
            str(e),
            context
        )
    
    # Handle different HTTP methods
    if http_method == 'GET':
        # List GameLift fleets
        return handle_list_fleets(gamelift_client, context)
        
    elif http_method == 'POST':
        # Handle other GameLift operations based on body
        body = event.get('body', '{}')
        try:
            body_data = json.loads(body) if isinstance(body, str) else body
        except json.JSONDecodeError:
            body_data = {}
            
        action = body_data.get('action', 'list_fleets')
        
        if action == 'list_fleets':
            return handle_list_fleets(gamelift_client, context)
        elif action == 'describe_fleet':
            fleet_id = body_data.get('fleet_id')
            if not fleet_id:
                return create_error_response(
                    400,
                    "Missing required parameter: fleet_id",
                    None,
                    context
                )
            return handle_describe_fleet(gamelift_client, fleet_id, context)
        else:
            return create_error_response(
                400,
                f"Unknown action: {action}",
                None,
                context
            )
    else:
        return create_error_response(
            405,
            f'Method {http_method} not supported',
            None,
            context
        )


def handle_list_fleets(gamelift_client: boto3.client, context: Any) -> Dict[str, Any]:
    """
    Handle ListFleets request
    
    Args:
        gamelift_client: Boto3 GameLift client
        context: Lambda context
        
    Returns:
        API Gateway response with fleet list
    """
    try:
        # Call GameLift ListFleets API
        response = gamelift_client.list_fleets()
        
        fleet_ids = response.get('FleetIds', [])
        next_token = response.get('NextToken', None)
        
        # Prepare response
        response_body = {
            'status': 'success',
            'operation': 'list_fleets',
            'fleet_count': len(fleet_ids),
            'fleets': fleet_ids,
            'next_token': next_token,
            'timestamp': context.aws_request_id if hasattr(context, 'aws_request_id') else 'N/A'
        }
        
        return {
            'statusCode': 200,
            'headers': {
                'Content-Type': 'application/json',
                'Access-Control-Allow-Origin': '*',
                'Access-Control-Allow-Headers': 'Content-Type,X-Amz-Date,Authorization,X-Api-Key',
                'Access-Control-Allow-Methods': 'GET,POST,OPTIONS'
            },
            'body': json.dumps(response_body, indent=2, default=str)
        }
        
    except ClientError as e:
        error_code = e.response['Error']['Code']
        error_message = e.response['Error']['Message']
        
        return create_error_response(
            500,
            f'GameLift API error: {error_code}',
            error_message,
            context
        )
    except Exception as e:
        return create_error_response(
            500,
            'Unexpected error listing fleets',
            str(e),
            context
        )


def handle_describe_fleet(gamelift_client: boto3.client, fleet_id: str, context: Any) -> Dict[str, Any]:
    """
    Handle DescribeFleet request
    
    Args:
        gamelift_client: Boto3 GameLift client
        fleet_id: Fleet ID to describe
        context: Lambda context
        
    Returns:
        API Gateway response with fleet details
    """
    try:
        # Call GameLift DescribeFleetAttributes API
        response = gamelift_client.describe_fleet_attributes(
            FleetIds=[fleet_id]
        )
        
        fleet_attributes = response.get('FleetAttributes', [])
        
        if not fleet_attributes:
            return create_error_response(
                404,
                f'Fleet not found: {fleet_id}',
                None,
                context
            )
        
        # Convert datetime objects to strings for JSON serialization
        fleet = fleet_attributes[0]
        
        response_body = {
            'status': 'success',
            'operation': 'describe_fleet',
            'fleet': convert_fleet_to_dict(fleet),
            'timestamp': context.aws_request_id if hasattr(context, 'aws_request_id') else 'N/A'
        }
        
        return {
            'statusCode': 200,
            'headers': {
                'Content-Type': 'application/json',
                'Access-Control-Allow-Origin': '*',
                'Access-Control-Allow-Headers': 'Content-Type,X-Amz-Date,Authorization,X-Api-Key',
                'Access-Control-Allow-Methods': 'GET,POST,OPTIONS'
            },
            'body': json.dumps(response_body, indent=2, default=str)
        }
        
    except ClientError as e:
        error_code = e.response['Error']['Code']
        error_message = e.response['Error']['Message']
        
        return create_error_response(
            500,
            f'GameLift API error: {error_code}',
            error_message,
            context
        )
    except Exception as e:
        return create_error_response(
            500,
            'Unexpected error describing fleet',
            str(e),
            context
        )


def convert_fleet_to_dict(fleet: Any) -> Dict[str, Any]:
    """
    Convert GameLift fleet object to dictionary for JSON serialization
    
    Args:
        fleet: GameLift fleet object
        
    Returns:
        Dictionary representation
    """
    return {
        'FleetId': fleet.get('FleetId'),
        'FleetArn': fleet.get('FleetArn'),
        'FleetType': fleet.get('FleetType'),
        'EC2InstanceType': fleet.get('EC2InstanceType'),
        'BuildId': fleet.get('BuildId'),
        'Status': fleet.get('Status'),
        'Description': fleet.get('Description'),
        'Name': fleet.get('Name'),
        'CreationTime': fleet.get('CreationTime').isoformat() if fleet.get('CreationTime') else None,
        'TerminationTime': fleet.get('TerminationTime').isoformat() if fleet.get('TerminationTime') else None,
    }


def create_error_response(
    status_code: int,
    message: str,
    details: Optional[str],
    context: Any
) -> Dict[str, Any]:
    """
    Create standardized error response
    
    Args:
        status_code: HTTP status code
        message: Error message
        details: Optional error details
        context: Lambda context
        
    Returns:
        API Gateway error response
    """
    error_body = {
        'status': 'error',
        'message': message,
        'timestamp': context.aws_request_id if hasattr(context, 'aws_request_id') else 'N/A'
    }
    
    if details:
        error_body['details'] = details
    
    return {
        'statusCode': status_code,
        'headers': {
            'Content-Type': 'application/json',
            'Access-Control-Allow-Origin': '*',
            'Access-Control-Allow-Headers': 'Content-Type,X-Amz-Date,Authorization,X-Api-Key',
            'Access-Control-Allow-Methods': 'GET,POST,OPTIONS'
        },
        'body': json.dumps(error_body, indent=2)
    }

File: modules/lambda/go_gamelift/src/gamelift_handler.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"os"

	"github.com/aws/aws-lambda-go/events"
	"github.com/aws/aws-lambda-go/lambda"
	"github.com/aws/aws-sdk-go-v2/config"
	"github.com/aws/aws-sdk-go-v2/service/gamelift"
	"github.com/aws/aws-sdk-go-v2/service/gamelift/types"
)

// GameLiftLambdaHandler handles Lambda requests
type GameLiftLambdaHandler struct {
	gameliftClient *gamelift.Client
}

// GameLiftResponse represents the API response structure
type GameLiftResponse struct {
	Status      string                 `json:"status"`
	Message     string                 `json:"message,omitempty"`
	Operation   string                 `json:"operation,omitempty"`
	FleetCount  int                    `json:"fleet_count,omitempty"`
	Fleets      []string               `json:"fleets,omitempty"`
	Fleet       *FleetInfo             `json:"fleet,omitempty"`
	NextToken   *string                `json:"next_token,omitempty"`
	Error       *ErrorInfo             `json:"error,omitempty"`
	Timestamp   string                 `json:"timestamp,omitempty"`
}

// FleetInfo represents fleet information
type FleetInfo struct {
	FleetId          *string `json:"FleetId"`
	FleetArn         *string `json:"FleetArn"`
	FleetType        types.FleetType `json:"FleetType"`
	EC2InstanceType  types.EC2InstanceType `json:"EC2InstanceType"`
	BuildId          *string `json:"BuildId"`
	Status           types.FleetStatus `json:"Status"`
	Description      *string `json:"Description"`
	Name             *string `json:"Name"`
	CreationTime     *string `json:"CreationTime,omitempty"`
	TerminationTime  *string `json:"TerminationTime,omitempty"`
}

// ErrorInfo represents error information
type ErrorInfo struct {
	Code    string `json:"code"`
	Message string `json:"message"`
}

// GameLiftRequest represents the incoming request
type GameLiftRequest struct {
	Action  string `json:"action"`
	FleetID string `json:"fleet_id"`
}

// Handler is the main Lambda handler function for GameLift operations
func (h *GameLiftLambdaHandler) Handler(ctx context.Context, request events.APIGatewayProxyRequest) (events.APIGatewayProxyResponse, error) {
	// Get request details
	httpMethod := request.HTTPMethod

	// Handle different HTTP methods
	switch httpMethod {
	case "GET":
		return h.handleListFleets(ctx, request)

	case "POST":
		// Parse request body
		var bodyData GameLiftRequest
		if err := json.Unmarshal([]byte(request.Body), &bodyData); err != nil {
			bodyData = GameLiftRequest{Action: "list_fleets"}
		}

		switch bodyData.Action {
		case "list_fleets":
			return h.handleListFleets(ctx, request)
		case "describe_fleet":
			if bodyData.FleetID == "" {
				return h.createErrorResponse(400, "Missing required parameter: fleet_id", "", request.RequestContext.RequestID), nil
			}
			return h.handleDescribeFleet(ctx, bodyData.FleetID, request)
		default:
			return h.createErrorResponse(400, fmt.Sprintf("Unknown action: %s", bodyData.Action), "", request.RequestContext.RequestID), nil
		}

	default:
		return h.createErrorResponse(405, fmt.Sprintf("Method %s not supported", httpMethod), "", request.RequestContext.RequestID), nil
	}
}

// handleListFleets handles ListFleets request
func (h *GameLiftLambdaHandler) handleListFleets(ctx context.Context, request events.APIGatewayProxyRequest) (events.APIGatewayProxyResponse, error) {
	// Call GameLift ListFleets API
	result, err := h.gameliftClient.ListFleets(ctx, &gamelift.ListFleetsInput{})
	if err != nil {
		return h.createErrorResponse(500, "Failed to list fleets", err.Error(), request.RequestContext.RequestID), nil
	}

	// Prepare response
	responseBody := GameLiftResponse{
		Status:     "success",
		Operation:  "list_fleets",
		FleetCount: len(result.FleetIds),
		Fleets:     result.FleetIds,
		NextToken:  result.NextToken,
		Timestamp:  request.RequestContext.RequestID,
	}

	// Marshal response to JSON
	responseJSON, err := json.Marshal(responseBody)
	if err != nil {
		return h.createErrorResponse(500, "Failed to encode response", err.Error(), request.RequestContext.RequestID), nil
	}

	// Return API Gateway response
	return events.APIGatewayProxyResponse{
		StatusCode: 200,
		Headers: map[string]string{
			"Content-Type":                 "application/json",
			"Access-Control-Allow-Origin":  "*",
			"Access-Control-Allow-Headers": "Content-Type,X-Amz-Date,Authorization,X-Api-Key",
			"Access-Control-Allow-Methods": "GET,POST,OPTIONS",
		},
		Body: string(responseJSON),
	}, nil
}

// handleDescribeFleet handles DescribeFleetAttributes request
func (h *GameLiftLambdaHandler) handleDescribeFleet(ctx context.Context, fleetID string, request events.APIGatewayProxyRequest) (events.APIGatewayProxyResponse, error) {
	// Call GameLift DescribeFleetAttributes API
	result, err := h.gameliftClient.DescribeFleetAttributes(ctx, &gamelift.DescribeFleetAttributesInput{
		FleetIds: []string{fleetID},
	})
	if err != nil {
		return h.createErrorResponse(500, "Failed to describe fleet", err.Error(), request.RequestContext.RequestID), nil
	}

	if len(result.FleetAttributes) == 0 {
		return h.createErrorResponse(404, fmt.Sprintf("Fleet not found: %s", fleetID), "", request.RequestContext.RequestID), nil
	}

	// Convert fleet to FleetInfo
	fleetAttribute := result.FleetAttributes[0]
	fleetInfo := &FleetInfo{
		FleetId:         fleetAttribute.FleetId,
		FleetArn:        fleetAttribute.FleetArn,
		FleetType:       fleetAttribute.FleetType,
		EC2InstanceType: fleetAttribute.EC2InstanceType,
		BuildId:         fleetAttribute.BuildId,
		Status:          fleetAttribute.Status,
		Description:     fleetAttribute.Description,
		Name:            fleetAttribute.Name,
	}

	// Format timestamps if available
	if fleetAttribute.CreationTime != nil {
		ct := fleetAttribute.CreationTime.Format("2006-01-02T15:04:05Z07:00")
		fleetInfo.CreationTime = &ct
	}
	if fleetAttribute.TerminationTime != nil {
		tt := fleetAttribute.TerminationTime.Format("2006-01-02T15:04:05Z07:00")
		fleetInfo.TerminationTime = &tt
	}

	// Prepare response
	responseBody := GameLiftResponse{
		Status:     "success",
		Operation:  "describe_fleet",
		Fleet:      fleetInfo,
		Timestamp:  request.RequestContext.RequestID,
	}

	// Marshal response to JSON
	responseJSON, err := json.Marshal(responseBody)
	if err != nil {
		return h.createErrorResponse(500, "Failed to encode response", err.Error(), request.RequestContext.RequestID), nil
	}

	// Return API Gateway response
	return events.APIGatewayProxyResponse{
		StatusCode: 200,
		Headers: map[string]string{
			"Content-Type":                 "application/json",
			"Access-Control-Allow-Origin":  "*",
			"Access-Control-Allow-Headers": "Content-Type,X-Amz-Date,Authorization,X-Api-Key",
			"Access-Control-Allow-Methods": "GET,POST,OPTIONS",
		},
		Body: string(responseJSON),
	}, nil
}

// createErrorResponse creates standardized error response
func (h *GameLiftLambdaHandler) createErrorResponse(statusCode int, message string, details string, timestamp string) events.APIGatewayProxyResponse {
	errorBody := GameLiftResponse{
		Status:    "error",
		Message:   message,
		Timestamp: timestamp,
	}

	if details != "" {
		errorBody.Error = &ErrorInfo{
			Code:    fmt.Sprintf("%d", statusCode),
			Message: details,
		}
	}

	responseJSON, _ := json.Marshal(errorBody)

	return events.APIGatewayProxyResponse{
		StatusCode: statusCode,
		Headers: map[string]string{
			"Content-Type":                 "application/json",
			"Access-Control-Allow-Origin":  "*",
			"Access-Control-Allow-Headers": "Content-Type,X-Amz-Date,Authorization,X-Api-Key",
			"Access-Control-Allow-Methods": "GET,POST,OPTIONS",
		},
		Body: string(responseJSON),
	}
}

// initGameLiftClient initializes the GameLift client
func initGameLiftClient() (*gamelift.Client, error) {
	// Load AWS config
	cfg, err := config.LoadDefaultConfig(context.TODO())
	if err != nil {
		return nil, fmt.Errorf("failed to load AWS config: %w", err)
	}

	// Create GameLift client
	client := gamelift.NewFromConfig(cfg)
	return client, nil
}

func main() {
	// Initialize GameLift client
	gameliftClient, err := initGameLiftClient()
	if err != nil {
		fmt.Printf("Error initializing GameLift client: %v\n", err)
		os.Exit(1)
	}

	// Create handler with GameLift client
	handler := &GameLiftLambdaHandler{
		gameliftClient: gameliftClient,
	}

	// Start Lambda handler
	lambda.Start(handler.Handler)
}

